qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -389,89 +525,169 @@ def _synthesize_operation(
+
+            if as_dag.num_qubits() != inner_context.num_qubits():
+                raise TranspilerError("HighLevelSynthesis internal error.")
+
+            # We save the current state of the tracker to be able to return the ancilla
+            # qubits to the current positions. Note that at this point we do not know
+            # which ancilla qubits will be allocated.
+            saved_tracker = tracker.copy()
+            synthesized = self._run(
+                as_dag, tracker, inner_context, use_ancillas=use_ancillas, top_level=False
+            )
+            synthesized_context = inner_context
+
+            if (synthesized is not None) and (synthesized.num_qubits() > len(qubits)):
+                # need to borrow more qubits from tracker
+                global_aux_qubits = tracker.borrow(
+                    synthesized.num_qubits() - len(qubits), context.to_globals(qubits)
+                )
+                global_to_local = context.to_local_mapping()
+
+                for aq in global_aux_qubits:
+                    if aq in global_to_local:
+                        qubits.append(global_to_local[aq])
+                    else:
+                        new_local_qubit = context.add_qubit(aq)
+                        qubits.append(new_local_qubit)
+
+            if len(qubits) > num_original_qubits:
+                tracker.replace_state(
+                    saved_tracker, context.to_globals(qubits[num_original_qubits:])
                 )
-                if re_synthesized is not None:
-                    synthesized = re_synthesized
-                used_qubits = qubits
-
-            elif isinstance(synthesized, QuantumCircuit):
-                aux_qubits = tracker.borrow(synthesized.num_qubits - len(qubits), qubits)
-                used_qubits = qubits + tuple(aux_qubits)
-                as_dag = circuit_to_dag(synthesized, copy_operations=False)
-
-                # map used qubits to subcircuit
-                new_qubits = [as_dag.find_bit(q).index for q in as_dag.qubits]
-                qubit_map = dict(zip(used_qubits, new_qubits))
-
-                synthesized = self._run(as_dag, tracker.copy(qubit_map))
-                if synthesized.num_qubits() != len(used_qubits):
-                    raise RuntimeError(
-                        f"Mismatching number of qubits, using {synthesized.num_qubits()} "
-                        f"but have {len(used_qubits)}."
-                    )
 
-            else:
-                raise RuntimeError(f"Unexpected synthesized type: {type(synthesized)}")
+        else:
+            raise TranspilerError(f"Unexpected synthesized type: {type(synthesized)}")
 
-        if synthesized is not None and used_qubits is None:
-            raise RuntimeError("Failed to find qubit indices on", synthesized)
+        if isinstance(synthesized, DAGCircuit) and synthesized_context is None:
+            raise TranspilerError("HighLevelSynthesis internal error.")
```

Comment:
We could be a little more specific about this error message if the reason for this not to work is known.

Reply:
Ok, I will try to see if I can think of better error messages... Or maybe I can differentiate these somehow, in one place say "something when horribly wrong", in another - "something went terribly wrong", etc.